k9mail-library/src/test/java/com/fsck/k9/mail/internet/DecoderUtilTest.java

```diff
@@ -109,5 +109,19 @@ public void testDecodeEncodedWords() {
         expect = "=?x?= =?";
         message = null;
         assertEquals(expect, DecoderUtil.decodeEncodedWords(body, message));
+
+        //Multi encoded header
+        body = "=?utf-8?B?5Liq5Lq66YKu566xOkJVRyAjMzAyNDY6OumCruS7tuato+aWh+mZhOS7tuWQ?=\n" +
+                "=?utf-8?B?jeensOecgeeVpeaYvuekuuS8mOWMlg==?=";
+        expect = "个人邮箱:BUG #30246::邮件正文附件��称省略显示优化";
```

Comment:
What we currently do might be technically correct because the text is split in the middle of an UTF-8 encoded character. But it's certainly not what users expect.
If you combine the two base64-encoded parts and only then UTF-8 decode we end up with: 个人邮箱:BUG #30246::邮件正文附件名称省略显示优化 (no replacement characters �; so presumably this is what the sender actually wrote)

Encoding this "wrong" behavior in a test doesn't seem useful.

Reply:
Hmm, if you copy it into translate.google.com it comes out having translated / skipped the unknowns and it's understandable.

The problem is the spec allows for mixed encoding - i.e. we can't guarantee we'll get another utf-8 string to follow I think.

What they are doing is definitely a standards violation:

> Each 'encoded-word' MUST encode an integral number of octets.  The
   'encoded-text' in each 'encoded-word' must be well-formed according
   to the encoding specified; the 'encoded-text' may not be continued in
   the next 'encoded-word'.

I don't mind too much trying to handle the case of breaking the spec (you're generally less keen on this than me :) ). We can probably just merge all sequential b64 streams of identical encoding given whitespace is ignored.

I was mostly concerned with checking we were actually decoding multiple encoded words in a single field and adding a test for the other encoding.

Reply:
> I was mostly concerned with checking we were actually decoding multiple encoded words in a single field and adding a test for the other encoding.

+1 for that. But let's use test data that is not broken.

If we keep the current test, we should probably add a comment saying it's fine to break the test if that gets us compatibility with broken clients. Decoding broken encoded words in a spec-compliant way is not an important property of our code that we want to enforce using tests.